### PR TITLE
Option for reporting GenTree operator bashing stats

### PR DESCRIPTION
**Pull request by pkukol**
This also adds a column to gtlist.h for "true" node sizes, BTW; @dotnet/jit-contrib PTAL.

### CODE REVIEW

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -394,6 +408,88 @@ bool GenTree::IsNodeProperlySized() const
...
 
+/*****************************************************************************
+ *
+ *  When 'NODEBASH_STATS' is enabled in "jit.h" we record all instances of
+ *  an existing GenTree node having its operator changed. This can be useful
+ *  for two (related) things - to see what is being bashed (and what isn't),
+ *  and to verify that the existing choices for what nodes are marked 'large'
+ *  are reasonable (to minimize "wasted" space).
+ *
+ *  And yes, the hash function / logic is simplistic, but it is conflict-free
```
Comment (pgavlin) [inline src/jit/gentree.cpp]: IMO we should at least file an issue to change this s.t. it uses one of the preexisting hash tables. I think any option will need some changes in order to be able to use `ICorJitHost;:allocateMemory` for allocation (rather than `Compiler::compGetMem`), but I think that's a better approach long term.

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -394,6 +408,88 @@ bool GenTree::IsNodeProperlySized() const
...
+    desc->bhCount   += 1;
+    desc->bhOperOld  = operOld;
+    desc->bhOperNew  = operNew;
+}
+
+void                GenTree::ReportOperBashing(FILE *f)
+{
+    unsigned        total = 0;
+
+    fflush(f);
```
Comment (pgavlin) [inline src/jit/gentree.cpp]: Why is this necessary?

File: src/jit/jit.h [hunk 3]
```diff
@@ -465,16 +465,17 @@ typedef ptrdiff_t ssize_t;
 #define MEASURE_NODE_SIZE 0   // Collect stats about GenTree node allocations.
 #define MEASURE_PTRTAB_SIZE 0 // Collect stats about GC pointer table allocations.
 #define EMITTER_STATS 0       // Collect stats on the emitter.
+#define NODEBASH_STATS 0      // Collect stats on changed gtOper values in GenTree's.
 
-#define VERBOSE_SIZES 0  // Always display GC info sizes. If set, DISPLAY_SIZES must also be set.
-#define VERBOSE_VERIFY 0 // Dump additional information when verifying code. Useful to debug verification bugs.
+#define VERBOSE_SIZES 0       // Always display GC info sizes. If set, DISPLAY_SIZES must also be set.
+#define VERBOSE_VERIFY 0      // Dump additional information when verifying code. Useful to debug verification bugs.
```
Comment (pgavlin) [inline src/jit/jit.h]: Was this alignment done manually, or is this from `jit-format`?

**Comment by pgavlin** [comment]
Note that the approach taken here will not catch all of the situations in which we change `gtOper`, since we frequently do so without calling `SetOper`. Is the information that this change will be able to collect sufficient?

**Comment by BruceForstall** [comment]
@pgavlin Is that true? I think all changes should go through SetOper(). Can you find an example?

**Comment by pgavlin** [comment]
@BruceForstall: lowering, rationalization, liveness, decomp, legacy codegen, and side-effect extraction all do this. The number of instances is quite a bit smaller that I thought, however: a simple grep for `gtOper[[:space:]]*[^=!]=` only returns ~20 results.

File: src/jit/gtlist.h [hunk 4]
```diff
@@ -9,255 +9,255 @@
...
-GTNODE(PUTARG_REG   , "putarg_reg"    ,0,GTK_UNOP)              // operator that places outgoing arg in register
-GTNODE(PUTARG_STK   , "putarg_stk"    ,0,GTK_UNOP)              // operator that places outgoing arg in stack
-GTNODE(RETURNTRAP   , "returnTrap"    ,0,GTK_UNOP|GTK_NOVALUE)  // a conditional call to wait on gc
-GTNODE(SWAP         , "swap"          ,0,GTK_BINOP|GTK_NOVALUE) // op1 and op2 swap (registers)
-GTNODE(IL_OFFSET    , "il_offset"     ,0,GTK_LEAF|GTK_NOVALUE)  // marks an IL offset for debugging purposes
+GTNODE(REG_VAR          , "regVar"       ,GenTreeLclVar      ,0,GTK_LEAF|GTK_LOCAL)     // register variable
+GTNODE(CLS_VAR          , "clsVar"       ,GenTreeClsVar      ,0,GTK_LEAF)               // static data member
+GTNODE(CLS_VAR_ADDR     , "&clsVar"      ,GenTreeClsVar      ,0,GTK_LEAF)               // static data member address
+GTNODE(STORE_CLS_VAR    , "st.clsVar"    ,char               ,0,GTK_LEAF|GTK_NOVALUE)   // store to static data member
+GTNODE(ARGPLACE         , "argPlace"     ,GenTreeArgPlace    ,0,GTK_LEAF)               // placeholder for a register arg
```
Comment (BruceForstall) [inline src/jit/gtlist.h]: Looks like this node is dead code that can be removed?

Comment (pgavlin) [inline src/jit/gtlist.h]: Which node? `GT_ARGPLACE` is still used...

File: src/jit/gentree.cpp (hunk 1 above)
Comment (pkukol) [inline src/jit/gentree.cpp]: I didn't necessarily mean to stick this in as is (sorry I didn't mention this sooner) - I did try to find a low-overhead pre-existing hash-like thing but failed (everything I tried needed a bunch of additional #include's and who knows what else would be dragged in). Can you please recommend the right hash / hashmap / whatever for this? It's a trivial [op1,op2] -> count mapping (basically).

Comment (BruceForstall) [inline src/jit/gentree.cpp]: SimplerHashTable is widely used.

---
In reply to: [78828482](https://github.com/dotnet/coreclr/pull/7152#discussion_r78828482) [](ancestors = 78828482)

File: src/jit/compiler.hpp [hunk 5]
```diff
@@ -1289,6 +1289,10 @@ inline void GenTree::SetOper(genTreeOps oper, ValueNumberUpdate vnUpdate)
 
     assert(GenTree::s_gtNodeSizes[oper] == TREE_NODE_SZ_SMALL || (gtDebugFlags & GTF_DEBUG_NODE_LARGE));
 
+#if NODEBASH_STATS
+    RecordOperBashing(OperGet(), oper);
```
Comment (BruceForstall) [inline src/jit/compiler.hpp]: don't actually need this `#if` the way you've defined RecordOperBashing (and you don't have it in CopyFrom, below)

File: src/jit/rationalize.cpp [hunk 6]
```diff
@@ -605,6 +605,9 @@ void Rationalizer::RewriteAssignment(LIR::Use& use)
             }
             JITDUMP("Rewriting GT_ASG(%s(X), Y) to %s(X,Y):\n", GenTree::NodeName(location->gtOper),
                     GenTree::NodeName(storeOper));
+#if NODEBASH_STATS
+            GenTree::RecordOperBashing(storeBlk->OperGet(), storeOper);
+#endif
             storeBlk->gtOper = storeOper;
             storeBlk->gtFlags &= ~GTF_DONT_CSE;
```
Comment (BruceForstall) [inline src/jit/rationalize.cpp]: Why don't we just change this to SetOper?

File: src/jit/gentree.cpp (hunk 2 above)
Comment (pkukol) [inline src/jit/gentree.cpp]: I'm glad you asked - there are several components that spew stuff to stdout and the output sometimes gets intermingled. It's always just debug spew (i.e. a "real" user would never see any of it) and it only happens when you build with several off-by-default / apparently never used #ifdef's enabled, and so one could argue that things are "good enough" as is (with the flush the mixing happens a lot less and, frankly, I didn't dig deeper into it) but if folks think this is worth investigating further let's definitely do so.

File: src/jit/rationalize.cpp (hunk 6 above)
Comment (pgavlin) [inline src/jit/rationalize.cpp]: Yeah, we should at least fix the few occurrences of direct writes to `gtOper` with a follow-up PR.

File: src/jit/jit.h (hunk 3 above)
Comment (pkukol) [inline src/jit/jit.h]: I fixed it manually under the assumption that it was the formatting thing that accidentally messed it up.

**Comment by BruceForstall** [comment]
Seems like they should all be changed to use SetOper(), or at least ChangeOperUnchecked() (btw, you should probably record this one, also)

---
In reply to: [247136494](https://github.com/dotnet/coreclr/pull/7152#issuecomment-247136494) [](ancestors = 247136494)

File: src/jit/gtlist.h (hunk 4 above)
Comment (BruceForstall) [inline src/jit/gtlist.h]: STORE_CLS_VAR

**Comment by pkukol** [comment]
As for catching all the spots where we bash operators, I discovered a few more as I was extracting just the changes for this PR from my zillions of local changes, and I'm sure I managed to leave some out (the diff's are somewhat hard to manage plus some new ones actually appeared between the time I started adding the reporting code and now), so let me take another pass over the code base and add any spots that are not reporting their bashings. And yes, the ones in the PR do normally catch virtually everything (and the counts are the main thing to look at when e.g. deciding how to size 'small' vs. 'large' nodes for maximum savings) but the new code is also supposed to flag *all* combinations ever seen (and my next PR will contain logic to detect any errors when doing this, BTW) so that one can verify that no illegal bashings ever happen (in both retail and debug, when the right #ifdef is enabled).

File: src/jit/rationalize.cpp (hunk 6 above)
Comment (pkukol) [inline src/jit/rationalize.cpp]: I have no idea - I didn't want to mess with any existing code whenever possible, I could see no reason why SetOper wasn't being used, but then SetOper has been around for over 20 years so someone presumably had to have a reason not to call it?

**Comment by BruceForstall** [comment]
Note that I think you should change all bashings to use a common code path (perhaps SetOper or ChangeOperUnchecked), instead of dropping calls to RecordOperBashing() all around. It wasn't clear what you were intending to do based on your last comment.

File: src/jit/compiler.hpp (hunk 5 above)
Comment (pkukol) [inline src/jit/compiler.hpp]: Ah, yes, my schizophrenia showing - I couldn't decide which was cleaner / less ugly so I used both; sorry about that. Any thoughts as far as which is better? The #if/#ifdef frequency is pretty high already but making calls disappear through macros or empty inline's isn't exactly the cleanest thing either. Opinions?

File: src/jit/rationalize.cpp (hunk 6 above)
Comment (BruceForstall) [inline src/jit/rationalize.cpp]: Perhaps there was a reason, but more likely someone was lazy, forgot, or didn't know.

File: src/jit/gtlist.h (hunk 4 above)
Comment (pkukol) [inline src/jit/gtlist.h]: I don't know, I asked about several times but nobody ever replied. I guess that means it's OK to just delete it (i.e. it's not meant for some future thing or whatever)?

File: src/jit/compiler.hpp (hunk 5 above)
Comment (BruceForstall) [inline src/jit/compiler.hpp]: I vote no `#ifdef` at call site.

File: src/jit/gtlist.h (hunk 4 above)
Comment (pkukol) [inline src/jit/gtlist.h]: I'm sorry but could you please clarify what exactly about ARGPLACE is wrong (or whatever)?

**Comment by pkukol** [comment]
Well to be honest I didn't expect so many places to bash operators without calling SetOper and I assumed that they were doing so for some (unknown to me) reason. When I started adding this (weeks ago) I only found like 3 spots but since then it sort of grew and I should have gone back and revisited the sprinkling of the reporting calls all over. Will do a new complete pass over stuff and make this happen in SetOper or its moral equivalent. If that doesn't reasonable please let me know. Thanks!

File: src/jit/gentree.cpp (hunk 1 above)
Comment (pkukol) [inline src/jit/gentree.cpp]: Oh, BTW, memory alloc was indeed the main issue I ran into when I tried to use one of the existing hash classes; since this code is strictly for an occasional quick check by a developer and should never be enabled in any "real" build, I simply took this shortcut. Of course that same argument also says that we might as well use whatever you guys think is cleanest (since it won't matter for real use anyway), so what do you recommend?

Comment (BruceForstall) [inline src/jit/gentree.cpp]: I suggest using SimplerHashTable because it's commonly used, and should "just work".

File: src/jit/gtlist.h (hunk 4 above)
Comment (pkukol) [inline src/jit/gtlist.h]: OK, so STORE_CLS_VAR needs to go; will do.

File: src/jit/compiler.hpp (hunk 5 above)
Comment (pkukol) [inline src/jit/compiler.hpp]: Agreed. Next commit.

File: src/jit/rationalize.cpp (hunk 6 above)
Comment (pkukol) [inline src/jit/rationalize.cpp]: Actually SetOper now does a whole bunch of stuff it didn't do originally so there are probably good reasons not to call it in places; I'll just change those to ChangeOperUnchecked() I think.

File: src/jit/gentree.cpp (hunk 1 above)
Comment (pkukol) [inline src/jit/gentree.cpp]: That sounds familiar (and I think I tried that already), but I'll take another stab at it.

**Comment by pkukol** [comment]
@dotnet-bot test OSX x64 Checked Build and Test

**Comment by pkukol** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build

**Comment by pkukol** [comment]
@dotnet-bot test Linux ARM Emulator Cross Release Build

**Comment by pgavlin** [comment]
LGTM. Do we have an issue open to track using a different implementation for the hash table?

**Comment by pkukol** [comment]
It's issue #7218. Thanks for everyone's help on this.
